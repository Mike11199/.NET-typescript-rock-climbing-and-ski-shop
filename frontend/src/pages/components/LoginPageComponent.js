import { Container, Row, Col, Form, Button} from "react-bootstrap";
import { useState } from "react";
import { Link } from "react-router-dom";
import Spinner from "react-bootstrap/Spinner";
import GoogleLoginButton from "components/GoogleLogIn";
import toast, { Toaster } from 'react-hot-toast';
import IceClimbingPhoto from "images/ski_mountaineering_5.png"
import IceCavePhoto from "images/ice_cave_2.png"
import "mobileStyles.css"


const LoginPageComponent = ({ loginUserApiRequest, reduxDispatch, setReduxUserState, googleLogin }) => {

  //react state for form - if validated or not
  const [validated, setValidated] = useState(false)  
  const [LogInUserResponseState, setLogInUserResponseState] = useState(
      {success:"", error:"", loading: false}
  )

  //event handler for form submission
  const handleSubmit = async (event) => {
    
    event.preventDefault()
    event.stopPropagation()
    
    const form = event.currentTarget.elements
    const email = form.email.value
    const password = form.password.value
    const doNotLogout = form.doNotLogout.checked


    if ( email === '' || password === ''){
      toast.dismiss();  //https://react-hot-toast.com/docs/toast
      toast.error('Please provide all form values.',{style: {borderRadius: '10px',background: '#333',color: '#fff'}})
      return
    }

    // If form is valid, send request to back end
    if (event.currentTarget.checkValidity() === true && email && password) {

        setLogInUserResponseState({loading: true})

        try {
          const res = await loginUserApiRequest(email, password, doNotLogout)
          console.log(res)
          
          setLogInUserResponseState({success: res.success, loading: false, error: ""})
          setValidated(true)

          // Dispatch the setReduxUserState action that is passed from the LoginPage.js component
          // The action is passed to the userReducer to update global state with the the user data
          if (res.userLoggedIn) {
            reduxDispatch(setReduxUserState(res.userLoggedIn))
          }

          toast.dismiss();
          toast.success('Logging you in!',{style: {borderRadius: '10px', background: '#333', color: '#fff',},})          

          // redirect to user or admin page depending on if user has admin privileges
          if (res.success === "user logged in" && !res.userLoggedIn.isAdmin){
            setTimeout(function() {window.location.assign('/user')}, 1000)     
          }else{
            setTimeout(function() {window.location.assign('/admin/orders')}, 1000)
          }
          return          
        } 
        catch (er) {
          const errorMessage = er?.response?.data?.message ?? 'error'
          setLogInUserResponseState({ error: errorMessage })
          toast.dismiss();
          toast.error('Error logging in!  Wrong credentials.', { style: {borderRadius: '10px',background: '#333',color: '#fff',},})
          

          // shake password box 
          const formControl = document.querySelector('#formBasicPassword');
          formControl.classList.add('shake')
          formControl.addEventListener('animationend', () => {
              formControl.classList.remove('shake');
          })
          return
        }
    }

    // If form is not valid, run shake animation and exit
    toast.dismiss()
    toast.error('Error logging in!  Wrong credentials.',{style: {borderRadius: '10px', background: '#333',color: '#fff'},})

    // shake password box 
    const formControl = document.querySelector('#formBasicPassword');
    formControl.classList.add('shake')
    formControl.addEventListener('animationend', () => {
    formControl.classList.remove('shake')
    })

    return
  }



  const displaySpinner = () =>{
    if(LogInUserResponseState && LogInUserResponseState.loading === true){
      return <Spinner as="span" animation="border" size="sm" role="status" aria-hidden="true"/>
    }
    else{
      return ""
    }
  }

  return (
    <>
    <Toaster/>
    <img className="ski_image" alt="ice_climbing_photo" src={IceClimbingPhoto} ></img>
    <div style={{display:"flex", height:"100%"}}>
    <img className="ice_cave_image" alt="ice_climbing_photo" src={IceCavePhoto} ></img>
    </div>
    <div style={{display:"flex", height:"100%"}}>
    <Container style={{maxWidth:"100%", marginTop:"5%" }} >
      <Row className="mt-5">
        <Col md={5} ></Col>
        <Col md={3} id="log_in_page_form_1" className="pr-md-2" style={{maxWidth:"100%", marginRight:"2%" }} >
          <h1>Login</h1>
          {/* Initial state of form validation is False */}
          <Form noValidate validated={validated} onSubmit={handleSubmit}>

            {/* Email */}
            <Form.Group className="mb-3" controlId="formBasicEmail">
              <Form.Label>Email address</Form.Label>
              <Form.Control
                name="email"
                required
                type="email"
                placeholder="Enter email"
              />
            </Form.Group>

            {/* Password */}
            <Form.Group className="mb-3" controlId="formBasicPassword" id="formBasicPassword">
              <Form.Label>Password</Form.Label>
              <Form.Control
                name="password"
                required
                type="password"
                placeholder="Password"
              />
            </Form.Group>

            {/* Checkbox to extend JWT to 7 days */}
            <Form.Group className="mb-3" controlId="formBasicCheckbox" >
              <Form.Check
                name="doNotLogout"
                type="checkbox"
                label="Keep me logged in."
                defaultChecked
              />
            </Form.Group>

            {/* Link to Register Instead */}
            <Row className="pb-2">
              <Col>
                Don't have an account? &nbsp; ðŸ‘‰ &nbsp;
                <Link to={"/register"}><strong>Register</strong></Link>
              </Col>
            </Row>

            {/* Submit Button */}
            <div width="300px">
            <Button variant="primary" type="submit"  style={{width:'230px'}}>
            {displaySpinner()}
              Login
            </Button>
            <GoogleLoginButton googleLogin={googleLogin} reduxDispatch={reduxDispatch}/>
            </div>

          </Form>
        </Col>
      </Row>
    </Container>
    </div>
    </>
  )
}

export default LoginPageComponent
