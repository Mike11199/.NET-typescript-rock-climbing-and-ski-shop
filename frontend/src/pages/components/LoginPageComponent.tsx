import { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import Spinner from "react-bootstrap/Spinner";
import toast, { Toaster } from "react-hot-toast";
import IceClimbingPhoto from "../../images/ski_mountaineering_5.png";
import IceCavePhoto from "../../images/ice_cave_2.png";
import "../../../src/mobileStyles.css";
import { AxiosResponse } from "axios";
import GoogleLoginButton from "../../../src/components/GoogleLogIn";
import { Button } from "react-bootstrap";
import { useWindowWidth } from "@react-hook/window-size";

interface LoginPageComponentProps {
  loginUserApiRequest: (
    email: string,
    password: string,
    doNotLogout: boolean
  ) => Promise<any>;
  reduxDispatch?: Function;
  setReduxUserState: Function;
  googleLogin?: (googleToken: string) => Promise<AxiosResponse<any>>;
}

interface LoginUserResponse {
  success: string;
  error: string;
  loading: boolean;
}

const LoginPageComponent = ({
  loginUserApiRequest,
  reduxDispatch,
  setReduxUserState,
  googleLogin,
}: LoginPageComponentProps) => {
  const [LogInUserResponseState, setLogInUserResponseState] =
    useState<LoginUserResponse>({
      success: "",
      error: "",
      loading: false,
    });
  const [loginWithDefaultUser, setLoginWithDefaultUser] = useState(false);

  const [formData, setFormData] = useState({
    email: "",
    password: "",
    doNotLogout: false,
  });

  const handleChange = (event) => {
    const { name, value, type, checked } = event.target;
    setFormData({
      ...formData,
      [name]: type === "checkbox" ? checked : value,
    });
  };

  //event handler for login form submission
  const handleSubmit = async (event) => {
    event.preventDefault();
    event.stopPropagation();

    let form = event.currentTarget.elements;
    let email = form.email.value;
    let password = form.password.value;
    let doNotLogout = form.doNotLogout.checked;

    if (loginWithDefaultUser) {
      email = "test@test.com";
      password = "easy_to_guess_password"; // this doesn't matter - API won't allow test user edits
    }

    if (email === "" || password === "") {
      toastError("Please provide all form values!");
      return;
    }

    setLogInUserResponseState((prevState) => ({
      ...prevState,
      loading: true,
    }));

    try {
      const res = await loginUserApiRequest(email, password, doNotLogout);
      if (!res?.userLoggedIn) throw Error;

      setLogInUserResponseState({
        success: res?.success,
        loading: false,
        error: "",
      });
      reduxDispatch!(setReduxUserState(res?.userLoggedIn));
      toastSuccess("Logging you in!");
      return redirectUserOrAdminAfterLogin(res);
    } catch (er: any) {
      setLogInUserResponseState((prevState) => ({
        ...prevState,
        error: er?.response?.data?.message ?? "error",
      }));
      setLogInUserResponseState((prevState) => ({
        ...prevState,
        loading: false,
      }));
      toastError("Error logging in!  Wrong credentials.");
      return shakeDivById("formBasicPassword");
    }
  };

  const displaySpinner = () => {
    if (LogInUserResponseState && LogInUserResponseState.loading === true) {
      return (
        <Spinner
          as="span"
          animation="border"
          size="sm"
          role="status"
          aria-hidden="true"
        />
      );
    } else {
      return "";
    }
  };

  const windowWidth = useWindowWidth();
  const isMobileView = windowWidth <= 600;

  return (
    <>
      <Toaster />
      <img
        className="ski_image"
        alt="ice_climbing_photo"
        src={IceClimbingPhoto}
      ></img>
      <div style={{ display: "flex", height: "100%" }}>
        <img
          className="ice_cave_image"
          alt="ice_climbing_photo"
          src={IceCavePhoto}
        />
      </div>

      <div className="login-container">
        <form onSubmit={handleSubmit} className="login-form">
          <h1>Login</h1>
          <div className="form-group">
            <label>Email address</label>
            <input
              name="email"
              placeholder="Enter email"
              value={formData.email}
              onChange={handleChange}
            />
          </div>
          <div className="form-group">
            <label>Password</label>
            <input
              id="formBasicPassword"
              name="password"
              type="password"
              placeholder="Password"
              value={formData.password}
              onChange={handleChange}
            />
          </div>
          <div>
            <div style={{ display: "flex", gap: "1rem" }}>
              <input
                name="doNotLogout"
                type="checkbox"
                checked={formData.doNotLogout}
                onChange={handleChange}
              />
              <label>Keep me logged in</label>
            </div>
            <div className="links">
              Don't have an account? ðŸ‘‰{" "}
              <strong>
                <Link to={"/register"}>Register</Link>
              </strong>
            </div>
          </div>
          <div className="button-group">
            <Button
              variant="primary"
              type="submit"
              style={{ width: isMobileView ? "320px" : "400px" }}
            >
              {displaySpinner()}
              Login
            </Button>
            <Button
              style={{ width: isMobileView ? "320px" : "400px" }}
              variant="danger"
              type="submit"
              onClick={(e) => {
                setLoginWithDefaultUser(true);
              }}
            >
              {displaySpinner()}
              Demo Login
            </Button>
            <GoogleLoginButton
              googleLogin={googleLogin}
              reduxDispatch={reduxDispatch}
            />
          </div>
        </form>
      </div>
    </>
  );
};

export default LoginPageComponent;

function toastError(text: string) {
  toast.dismiss();
  toast.error(text, {
    style: { borderRadius: "10px", background: "#333", color: "#fff" },
  });
}

function toastSuccess(text: string) {
  toast.dismiss();
  toast.success(text, {
    style: { borderRadius: "10px", background: "#333", color: "#fff" },
  });
}

/**
 * Redirects the user or admin to the appropriate page after a successful login.
 *
 * @param {any} res - The response object from the login API request.
 * @returns {void}
 */
function redirectUserOrAdminAfterLogin(res: any) {
  if (res?.success === "user logged in" && !res?.userLoggedIn?.isAdmin) {
    setTimeout(function () {
      window?.location?.assign("/user");
    }, 1000);
  } else {
    setTimeout(function () {
      window?.location?.assign("/admin/orders");
    }, 1000);
  }
}

function shakeDivById(id: string): void {
  const formControl = document?.querySelector(`#${id}`);
  if (formControl) {
    formControl?.classList.add("shake");
    formControl?.addEventListener("animationend", () => {
      formControl?.classList.remove("shake");
    });
  }
}
