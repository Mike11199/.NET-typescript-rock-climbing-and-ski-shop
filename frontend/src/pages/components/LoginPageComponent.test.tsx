// import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import LoginPageComponent from "./LoginPageComponent";
// import { BrowserRouter as Router } from "react-router-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer"; //for snapshot test

let loginUserApiRequest = () => {
  return new Promise((resolve, reject) => {
    resolve({
      success: "user logged in",
      userLoggedIn: {
        _id: "hd73hDsSDj",
        name: "John",
        lastName: "Doe",
        email: "admin@admin.com",
        isAdmin: true,
        doNotLogout: false,
      },
    });
  });
};

let loginUserApiRequestError = () => {
  return new Promise((resolve, reject) => {
    reject({
      response: {
        data: {
          message: "wrong credentials",
        },
      },
    });
  });
};

// Mock the document object
// const mockDocument = {
//   querySelector: jest.fn().mockReturnValue({
//     classList: {
//       add: jest.fn(),
//       remove: jest.fn(),
//     },
//     addEventListener: jest.fn(),
//   }),
// };

const assign = window.location.assign; //real assign function

// this runs before testing to set mock implementations of functions
beforeAll(() => {
  //create mock function for window.location.assign real function
  Object.defineProperty(window, "location", {
    value: { assign: jest.fn() },
  });

  // remove this as prevented code to shake css password box if node env is testing - shakes only works in prod now which uses querySelector

  // jest.spyOn(document, "querySelector").mockImplementation(() => ({
  //   classList: {
  //     add: jest.fn(),
  //     remove: jest.fn(),
  //   },
  //   addEventListener: jest.fn(),
  // }));

  //mock implementation of matchMedia function as was breaking when testing toast notifications with react-hot-toast - reference ChatGPT help
  window.matchMedia = jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
  }));
});

afterAll(() => {
  window.location.assign = assign; //after tests go back to normal
  jest.restoreAllMocks();
});

test("if admin is logged in", async () => {
  // const querySelectorSpy = jest.spyOn(document, "querySelector");
  // const addEventListenerSpy = jest.spyOn(document, "addEventListener");

  render(
    <Router>
      <LoginPageComponent
        loginUserApiRequest={loginUserApiRequest}
        reduxDispatch={() => {}}
        setReduxUserState={() => {}}
      />
    </Router>,
  );
  await screen.findByLabelText("Email address");
  expect(screen.getByLabelText("Email address")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Password")).toBeInTheDocument();
  expect(screen.getByText("Login", { selector: "button" })).toBeInTheDocument();

  const emailField = screen.getByLabelText("Email address");
  const passwordField = screen.getByPlaceholderText("Password");
  const submitButton = screen.getByText("Login", { selector: "button" });

  fireEvent.focus(emailField);
  fireEvent.change(emailField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(emailField);
  fireEvent.focus(passwordField);
  fireEvent.change(passwordField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(passwordField);

  fireEvent.click(submitButton);

  await screen.findByLabelText("Email address");
  // console.log(passwordField); // Add this line
  expect(window.location.assign).toHaveBeenCalledWith("/admin/orders");
});

test("if wrong credentials toast notification modification", async () => {
  render(
    <Router>
      <LoginPageComponent
        loginUserApiRequest={loginUserApiRequestError}
        reduxDispatch={() => {}}
        setReduxUserState={() => {}}
      />
    </Router>,
  );
  await screen.findByLabelText("Email address");
  expect(screen.getByLabelText("Email address")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Password")).toBeInTheDocument();
  expect(screen.getByText("Login", { selector: "button" })).toBeInTheDocument();

  const emailField = screen.getByLabelText("Email address");
  const passwordField = screen.getByPlaceholderText("Password");
  const submitButton = screen.getByText("Login", { selector: "button" });

  fireEvent.focus(emailField); //to place cursor on a field
  fireEvent.change(emailField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(emailField); //to click somewhere outside of field

  fireEvent.focus(passwordField);
  fireEvent.change(passwordField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(passwordField);

  fireEvent.click(submitButton);

  await screen.findByText(/Wrong credentials/i);
  // console.log('Found element:', await screen.findByText(/Wrong credentials/i));
  expect(screen.getByText(/Wrong credentials/i)).toBeInTheDocument();
});

// this test css and html
// documentation - https://jestjs.io/docs/snapshot-testing
test("create login snapshot", () => {
  // this creates a snapshot in the __snapshots__ folder with the html to test the look of the page
  // this could detect changing a button for example
  const tree = renderer
    .create(
      <Router>
        <LoginPageComponent
          loginUserApiRequest={loginUserApiRequest}
          reduxDispatch={() => {}}
          setReduxUserState={() => {}}
        />
      </Router>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
