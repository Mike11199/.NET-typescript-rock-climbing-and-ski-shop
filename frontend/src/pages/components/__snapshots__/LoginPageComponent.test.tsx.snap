// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create login snapshot 1`] = `
Array [
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  >
    <div
      className="go4109123758"
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translateY(0px)",
          "transition": undefined,
        }
      }
    >
      <div
        className="go2072408551"
        style={
          Object {
            "background": "#333",
            "borderRadius": "10px",
            "color": "#fff",
            "opacity": 0,
          }
        }
      >
        <div
          className="go685806154"
        >
          <div
            className="go1858758034"
          />
          <div
            className="go1579819456"
          >
            <div
              className="go2534082608"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          className="go3958317564"
          role="status"
        >
          Error logging in!  Wrong credentials.
        </div>
      </div>
    </div>
    <div
      className=""
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translateY(0px)",
          "transition": undefined,
        }
      }
    >
      <div
        className="go2072408551"
        style={
          Object {
            "background": "#333",
            "borderRadius": "10px",
            "color": "#fff",
            "opacity": 0,
          }
        }
      >
        <div
          className="go685806154"
        >
          <div
            className="go1858758034"
          />
          <div
            className="go1579819456"
          >
            <div
              className="go2344853693"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          className="go3958317564"
          role="status"
        >
          Logging you in!
        </div>
      </div>
    </div>
  </div>,
  <div
    className="container"
  >
    <div
      className="mt-5 justify-content-md-center row"
    >
      <div
        className="col-md-6"
      >
        <h1>
          Login
        </h1>
        <form
          className=""
          noValidate={true}
          onSubmit={[Function]}
        >
          <div
            className="mb-3"
          >
            <label
              className="form-label"
              htmlFor="formBasicEmail"
            >
              Email address
            </label>
            <input
              className="form-control"
              id="formBasicEmail"
              name="email"
              placeholder="Enter email"
              required={true}
              type="email"
            />
          </div>
          <div
            className="mb-3"
            id="formBasicPassword"
          >
            <label
              className="form-label"
              htmlFor="formBasicPassword"
            >
              Password
            </label>
            <input
              className="form-control"
              id="formBasicPassword"
              name="password"
              placeholder="Password"
              required={true}
              type="password"
            />
          </div>
          <div
            className="mb-3"
          >
            <div
              className="form-check"
            >
              <input
                className="form-check-input"
                disabled={false}
                id="formBasicCheckbox"
                name="doNotLogout"
                type="checkbox"
              />
              <label
                className="form-check-label"
                htmlFor="formBasicCheckbox"
                title=""
              >
                Do not logout
              </label>
            </div>
          </div>
          <div
            className="pb-2 row"
          >
            <div
              className="col"
            >
              Don't have an account? Â  ðŸ‘‰ Â 
              <a
                href="/register"
                onClick={[Function]}
              >
                <strong>
                  Register
                </strong>
              </a>
            </div>
          </div>
          <div
            width="300px"
          >
            <button
              className="btn btn-primary"
              disabled={false}
              style={
                Object {
                  "width": "230px",
                }
              }
              type="submit"
            >
              
              Login
            </button>
            <div
              style={
                Object {
                  "justifyContent": "center",
                  "marginTop": "20px",
                }
              }
            >
              <div
                style={
                  Object {
                    "height": 40,
                  }
                }
              />
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>,
]
`;
